File: START145C/F.cpp
#include <bits/stdc++.h>

#define ll long long int
#define yes cout << "YES" << endl
#define no cout << "NO" << endl

#define fastio()                      \
    ios_base::sync_with_stdio(false); \
    cin.tie(NULL);                    \
    cout.tie(NULL)

using namespace std;

///////////////////////////////////////////////////////
//                                                   //
//  Nickname: Atik                                   //
//  University: Green University of Bangladesh       //
//  Country: Bangladesh                              //
//  City: Narayanganj                                //
//  CodeForces: codeforces.com/profile/md.atik       //
//                                                   //
//  Language: C++                                    //
//  Compiler: gnu g++                                //
//  Editor: VS Code                                  //
//                                                   //
///////////////////////////////////////////////////////

/*int Vowel(string str)
{
    int len = str.length();
    int count = 0;
    for (int i = 0; i < len; i++)
    {
        if (str[i] == 'a' || str[i] == 'e' || str[i] == 'i' || str[i] == 'o' || str[i] == 'u' || str[i] == 'A' || str[i] == 'E' || str[i] == 'I' || str[i] == 'O' || str[i] == 'U')
            count++;
    }
    return count;
}

ll DIV(ll a, ll b)
{
    return (a + b - 1) / b;
}*/
const int MOD = 1e9 + 7;

ll factorial(int n)
{
    ll res = 1;
    for (int i = 2; i <= n; ++i)
    {
        res = (res * i) % MOD;
    }
    return res;
}

vector<int> isPrime(int n)
{
    vector<int> primes;
    vector<bool> is_prime(n + 1, true);
    is_prime[0] = is_prime[1] = false;

    for (int i = 2; i * i <= n; i++)
    {
        if (is_prime[i])
        {
            for (int j = i * i; j <= n; j += i)
            {
                is_prime[j] = false;
            }
        }
    }

    for (int i = 2; i <= n; i++)
    {
        if (is_prime[i])
        {
            primes.push_back(i);
        }
    }

    return primes;
}

long long mod_pow(long long base, long long exp, long long mod)
{
    long long result = 1;
    while (exp > 0)
    {
        if (exp % 2 == 1)
        {
            result = (result * base) % mod;
        }
        base = (base * base) % mod;
        exp /= 2;
    }
    return result;
}

vector<int> computePrefixSum(const vector<int> &nums)
{
    int n = nums.size();
    vector<int> prefixSum(n + 1, 0);

    for (int i = 1; i <= n; ++i)
    {
        prefixSum[i] = prefixSum[i - 1] + nums[i - 1];
    }

    return prefixSum;
}

const int MAXN = 110;
bool visited[MAXN];
vector<int> adj[MAXN];
void dfs(int v)
{
    visited[v] = true;
    for (int i = 0; i < adj[v].size(); i++)
    {
        int w = adj[v][i];
        if (!visited[w])
            dfs(w);
    }
}

void before_dfs()
{
    int n, m, x, y, sol;
    cin >> n >> m;
    for (int i = 0; i < m; i++)
    {
        cin >> x >> y;
        adj[x].push_back(y);
        adj[y].push_back(x);
    }

    for (int i = 1; i <= n; i++)
    {
        if (!visited[i])
        {
            dfs(i);
            sol++;
        }
    }

    cout << sol << endl;
}
const int MAX_N = 1e5 + 5;

vector<int> adj[MAX_N];
int G[MAX_N];
int subtree_gcd[MAX_N];

int gcd(int a, int b)
{
    if (b == 0)
        return a;
    return gcd(b, a % b);
}
void dfs3(int u, int parent)
{
    subtree_gcd[u] = u;
    G[u] = -1;

    for (int v : adj[u])
    {
        if (v != parent)
        {
            dfs3(v, u);
            if (v < u)
            {
                G[u] = max(G[u], gcd(u, subtree_gcd[v]));
            }
            subtree_gcd[u] = gcd(subtree_gcd[u], subtree_gcd[v]);
        }
    }
}
void MUKU()
{
    int N, R;
    cin >> N >> R;

    for (int i = 1; i <= N; i++)
    {
        adj[i].clear();
    }

    for (int i = 0; i < N - 1; i++)
    {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    dfs3(R, 0);

    for (int i = 1; i <= N; i++)
    {
        cout << G[i] << " ";
    }
    cout << "\n";
}

int main()
{
    fastio();

    int t;
    cin >> t;
    while (t--)
    {
        MUKU();
    }

    return 0;
}

File: 958_CF/F.py
#include <iostream>
#include <vector>

using namespace std;

const int MOD = 998244353;
const int MAXN = 701;

long long dp[MAXN][MAXN][MAXN];
long long fact[MAXN], inv_fact[MAXN];

long long pow_mod(long long base, long long exp) {
    long long result = 1;
    while (exp > 0) {
        if (exp & 1) result = (result * base) % MOD;
        base = (base * base) % MOD;
        exp >>= 1;
    }
    return result;
}

void precompute() {
    fact[0] = inv_fact[0] = 1;
    for (int i = 1; i < MAXN; i++) {
        fact[i] = (fact[i-1] * i) % MOD;
        inv_fact[i] = pow_mod(fact[i], MOD - 2);
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    precompute();

    int n;
    cin >> n;
    vector<int> a(n), b(n), c(n);
    for (int i = 0; i < n; i++) cin >> a[i];
    for (int i = 0; i < n; i++) cin >> b[i];
    for (int i = 0; i < n; i++) cin >> c[i];

    dp[1][1][1] = 1;
    for (int i = 2; i <= n; i++) {
        for (int j = 1; j <= i; j++) {
            for (int k = 1; k <= i; k++) {
                dp[i][j][k] = (dp[i-1][j-1][k] + dp[i-1][j][k-1]) % MOD;
                if (j < i && k < i) {
                    dp[i][j][k] = (dp[i][j][k] + dp[i-1][j][k] * (i - 2)) % MOD;
                }
            }
        }
    }

    vector<long long> result(n);
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= i; j++) {
            for (int k = 1; k <= i; k++) {
                for (int z = 0; z <= i - 1; z++) {
                    long long ways = (dp[i][j][k] * fact[i-1] % MOD) * inv_fact[z] % MOD * inv_fact[i-1-z] % MOD;
                    long long cost = (1LL * a[j-1] * b[k-1] % MOD) * c[z] % MOD;
                    result[i-1] = (result[i-1] + ways * cost) % MOD;
                }
            }
        }
    }

    for (int i = 0; i < n; i++) {
        cout << result[i] << " ";
    }
    cout << endl;

    return 0;
}